### PR TITLE
Generalize parameter verification and broadcast

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75977
* #75976
* __->__ #75970
* #75945

Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

### CODE REVIEW

File: torch/distributed/utils.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import torch.distributed as dist
+
+def _verify_param_shape_across_processes(process_group, tensors, logger=None):
+    return dist._verify_params_across_processes(process_group, tensors, logger)
+
+def _sync_params_and_buffers(
```
Comment (awgu) [inline torch/distributed/utils.py:6]: nit: Do we want to add some type annotations?

File: torch/distributed/utils.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import torch.distributed as dist
+
+def _verify_param_shape_across_processes(process_group, tensors, logger=None):
```
Comment (awgu) [inline torch/distributed/utils.py:3]: nit: `_verify_param_shapes_across_processes()` since verifying multiple parameters shapes?

**Review by awgu — APPROVED** [review]
Looks good to me! Just some nits.

**Review by zhaojuanmao — APPROVED** [review]
nice generalization!

**Comment by zhaojuanmao** [comment]
by the way, let's do benchmarks before landing as well just in case

File: torch/distributed/utils.py (hunk 1 above)
Comment (rohan-varma) [reply torch/distributed/utils.py:6]: I think we should probably decide whether we want to invest in typing across distributed codebase and take that on as a BE task?

**Pull request closed without merging**